### PR TITLE
[FLINK-7964] Add Apache Kafka 1.0/1.1 connectors

### PR DESCRIPTION
**Pull request by yanghua**
## What is the purpose of the change

*This pull request add Apache Kafka 1.0/1.1 connectors*

## Brief change log

  - *Add Apache Kafka 1.0/1.1 connectors*

## Verifying this change

This change added tests and can be verified as follows:

  - *FlinkKafka10ProducerITCase*
  - *FlinkKafkaProducer10ITCase*
  - *FlinkKafkaProducer10StateSerializerTest*
  - *Kafka10ITCase*
  - *Kafka10ProducerAtLeastOnceITCase*
  - *Kafka10ProducerExactlyOnceITCase*
  - *Kafka10TableSourceSinkFactoryTest*
  - *KafkaTestEnvironmentImpl*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by yanghua** [comment]
@tzulitai and @dawidwys @pnowojski Can someone help review this PR? thanks.

File: flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/Kafka10ITCase.java [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+import javax.annotation.Nullable;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.Optional;
+
+/**
+ * IT cases for Kafka 1.0 .
+ */
+public class Kafka10ITCase extends KafkaConsumerTestBase {
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/Kafka10ITCase.java]: Can you introduce base class for tests similar as it is done for producer? I mean when `@Test`s are defined in the base class, there is usually less boilerplate since you do not have to redefine the same test over and over again.

File: flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer10StateSerializerTest.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import org.apache.flink.streaming.api.functions.sink.TwoPhaseCommitSinkFunction;
+import org.apache.flink.streaming.api.functions.sink.TwoPhaseCommitSinkFunction.TransactionHolder;
+
+import java.util.Collections;
+import java.util.Optional;
+
+/**
+ * A test for the {@link TypeSerializer TypeSerializers} used for the Kafka producer state.
+ */
+public class FlinkKafkaProducer10StateSerializerTest
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer10StateSerializerTest.java]: Deduplicate this with `org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer011StateSerializerTest`? Introduce some base `FlinkKafkaProducerStateSerializerTestBase<TransactionState, TransactionContext>` generic class and instantiate it with either `FlinkKafkaProducer10` or `FlinkKafkaProducer011` classes?

File: flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestEnvironmentImpl.java [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+import scala.collection.mutable.ArraySeq;
+
+import static org.apache.flink.util.NetUtils.hostAndPortToUrlString;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * An implementation of the KafkaServerProvider for Kafka 1.0 .
+ */
+public class KafkaTestEnvironmentImpl extends KafkaTestEnvironment {
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestEnvironmentImpl.java]: This is either fully or largely duplicated with `org.apache.flink.streaming.connectors.kafka.KafkaTestEnvironmentImpl` from 0.11 connector.

File: flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer10ITCase.java [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+import static org.apache.flink.util.ExceptionUtils.findThrowable;
+import static org.apache.flink.util.Preconditions.checkState;
+import static org.hamcrest.Matchers.lessThan;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+
+/**
+ * IT cases for the {@link FlinkKafkaProducer10}.
+ */
+public class FlinkKafkaProducer10ITCase extends KafkaTestBase {
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer10ITCase.java]: Same here, please deduplicate with `org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer011ITCase`

File: flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafka10ProducerITCase.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import java.util.Properties;
+import java.util.UUID;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests for our own {@link FlinkKafka10Producer}.
+ */
+@SuppressWarnings("serial")
+public class FlinkKafka10ProducerITCase extends KafkaTestBase {
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafka10ProducerITCase.java]: Ditto, deduplicate with: `org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducerITCase`? Deduplication may require extracting some interface with `org.apache.flink.streaming.connectors.kafka.internal.FlinkKafkaProducer#resumeTransaction` method.

**Comment by yanghua** [comment]
@pnowojski this is a big PR, may need some discussion,  some questions : 

* I know that most of the code for some tests is repetitive, but I have to copy them because I need to modify the relevant statements that create objects, such as `FlinkKafkaConsumer011` -> `FlinkKafkaConsumer10`.
* At first I didn't know if I tried to keep the existing code implementation unchanged. Now I know that we can make some abstractions to reuse some code, and I will make some improvements.
* Do you think we need a single module named `flink-connector-kafka-1.1`?

**Comment by pnowojski** [comment]
1. I haven't thought this through, but one option would be by using generics. Other by extracting some common interfaces. With a base class of `FlinkKafkaConsumerBase`, you could probably use it in the tests with defining one abstract method `protected FlinkKafkaConsumerBase createFlinkKafkaConsumer(...)`.
2. I think we should refactor old tests and deduplicate the code. It's very depressing when basically one variable rename `sequenceNumbers` to `nextSequence ` is forcing us to add ~3000 lines of codes. 
3. I don't know, since I have no idea what's the status of 1.1 support by our `0.11` connector or this new `1.0` connector. If this newly added `1.0` connector can support without any issues Kafka `1.1`, there might be no need to add new module, but still this should be documented.

**Comment by yanghua** [comment]
@pnowojski I am refactoring the code to reduce unnecessary code. 

Actually, the change from kafka v0.11 to v1.0, Not just a variable, there are some tests code that are not available, see [here](https://github.com/apache/flink/pull/6577/files#diff-527838fe63568e530d871d45a98e3fe4R157) and [here](https://github.com/apache/flink/pull/6577/files#diff-527838fe63568e530d871d45a98e3fe4R168). Some [old methods about using Zookeeper to get metadata](https://github.com/apache/kafka/commit/1d24e10aeab616eede416201336e928b9a8efa98#commitcomment-30035522) which are not available.

for the third point, according @aljoscha 's [comment](https://issues.apache.org/jira/browse/FLINK-10067?focusedCommentId=16571628&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16571628), I think we also need a new connector for kafka v1.1 to categorize the problems encountered by different kafka versions.

**Comment by pnowojski** [comment]
Yes, I missed those changes.

Btw, don't get me wrong, with "depressing" I wasn't complaining about this contribution, since it's fully following already established pattern :) Also by me, since when I was creating `0.11` connector, I only refactored/deduplicated producer side of the tests and copied/pasted most of the consumer side tests.

Regarding the @aljoscha 's comment, I will clarify this, since the issues that he mentioned/referred was a little bit irrelevant to question whether we need another module if our client is forward compatible (if it's able to talk to newer Kafka servers).  FLINK-9690/FLINK-9700 were/are about incompatibilities of our `0.11` connector when compiling it against Kafka `1.0` or `1.1`. That's something else.

**Comment by yanghua** [comment]
@pnowojski  Ok, you are a gentleman and gave me a lot of good advice. I think it's necessary to create a new module for each large version of kafka, but if we create a separate module for each sub version of kafka, it does seem a bit more. But we did use some kafka non-public APIs, and I don't know if they have the possibility to be modified in the sub version. If you don't add a separate module for the sub version, you will only be able to use branch decisions like if/else in the future.

Are you working in the same place with @aljoscha ? If so, can you discuss this issue offline? And I will first refactor the current duplicate test code. Do you agree?

**Comment by pnowojski** [comment]
There is a risk that our reflection code will brake on sub version upgrade, however it's not that big a problem, since in our connectors we are freezing Kafka client's version that we are using. Keep in mind that Kafka client's version that we are using is independent of Kafka server's that we are talking to. For example our Kafka 0.11.3 connector as far as I know is able to work with both Kafka 1.0 and 1.1 (with lower performance however).

However, I thought a little bit more about arguments for providing separate module for each major Kafka version and there is one huge argument in favour of that: bug fixes in Kafka. There are quite a lot of bugs in Kafka and sometimes they are not backported to older versions. I think this reason is good enough for adding `flink-connector-kafka-1.1` module, even if our code is almost identical. Also this would reduce our user's confusion "Does flink-connector-kafka-0.11 works with Kafka 1.0?".

**Comment by yanghua** [comment]
@pnowojski Agree, then I will first refactor the duplicate code based on 1.0, then add the 1.1 module. In fact, 2.0 is also in the process of being prepared, but it is still on another branch of my locality.

File: flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/Kafka10ITCase.java (hunk 1 above)
Comment (yanghua) [reply flink-connectors/flink-connector-kafka-1.0/src/test/java/org/apache/flink/streaming/connectors/kafka/Kafka10ITCase.java]: @pnowojski I tried to extend from Kafka011ITCase : 

```java
public class Kafka10ITCase extends Kafka011ITCase
```

but the prepare method : 

```java
@BeforeClass
public static void prepare() throws ClassNotFoundException {
    KafkaProducerTestBase.prepare();    //here
    ((KafkaTestEnvironmentImpl) kafkaServer).setProducerSemantic(FlinkKafkaProducer011.Semantic.AT_LEAST_ONCE);
}
```
will trigger the method call chain : 

```java
protected static void startClusters(boolean secureMode, boolean hideKafkaBehindProxy) throws ClassNotFoundException {

    // dynamically load the implementation for the test
    //here will load KafkaTestEnvironmentImpl from connector 0.11, not 1.0
    Class<?> clazz = Class.forName("org.apache.flink.streaming.connectors.kafka.KafkaTestEnvironmentImpl");    
    kafkaServer = (KafkaTestEnvironment) InstantiationUtil.instantiate(clazz);

    LOG.info("Starting KafkaTestBase.prepare() for Kafka " + kafkaServer.getVersion());
```

the `startClusters` is a static class, the `clazz` is the instance of KafkaTestEnvironmentImpl from 0.11 module, and we can not use polymorphism to get a instance of KafkaTestEnvironmentImpl from 1.0.

**Comment by aljoscha** [comment]
@yanghua Sorry for commenting so late but I have a proposal.

I would like to take this chance and fix a shortcoming in the dependency structure of our connectors. The connector dependencies form a chain, i.e. `0.11` depends on `0.10`, which depends on `0.9`, which depends on base. This itself makes the dependencies weird and can lead to problems when packaging a connector as a jar (as @twalthr found out, maybe you can comment about that here). Also, the base connector module has a dependency on `kafka_2.11`, which it shouldn't need, only `kafka-clients` should be required.

I think we can split it such that we have a base kafka connector module and the modules for the individual versions only depend on that, not on connectors for specific previous kafka versions.

What do you think?

**Comment by aljoscha** [comment]
I think we would need an API call bridge (`KafkaConsumerBridge` and `KafkaProducerBridge`) and a factory for these (`KafkaBridgeFactory`). Those interfaces are defined in the base package, and they basically have a method for all the methods that we want to use on the consumer or producer. The source/sink (for example the Fetcher) are implemented against those interfaces. Specific version implementations then only need to provide implementations for those. Thus cleanly separating things. With this, we don't need any Kafka dependency in the base package.

**Comment by yanghua** [comment]
@aljoscha I agree with you. The current chain of dependencies does make refactoring difficult and not easy to abstract.

My question is that for different versions of kafka client, if the API has hardly changed (for example, kafka 0.11 and kafka 1.0), you are not in favor of flink-connector-kafka-1.0 inheriting the implementation of flink-connector-kafka-0.11? ? Or your idea is to extract some util methods (if the other connectors are independent of each other except the base module)? But these util methods may only work for both versions and not for all. If we don't take an inherited solution, it will result in a lot of duplicate code.

In addition, I have done some common implementation abstraction for all kafka connector test code, but I don't know why the 0.9 version of the connector can't pass the test. I plan to fall back to the first commit, then split multiple commits to change it, and push each commit to the PR to verify that it doesn't break the existing tests and simplify the complexity of the positioning problem.What do you think of this? 

cc @pnowojski

**Comment by eliaslevy** [comment]
What is the need for all this different Kafka connector versions?  Newer Kafka clients are compatible with older brokers.

**Comment by yanghua** [comment]
@eliaslevy Piotr's [comment](https://github.com/apache/flink/pull/6577#issuecomment-414964833) should answer your question.

**Comment by aljoscha** [comment]
@eliaslevy Is there a resource on the forward/backward compatibility of Kafka clients/brokers? I was under the impression that only from 1.0.0 onward can you properly use newer clients with older brokers. If that is the case things will get easier for us in the future and we only need the most recent versions.

@yanghua Yes, I'm in favour of connectors only depending on the base package, which has an abstraction API.

**Comment by pnowojski** [comment]
Good point @aljoscha with those dependencies, it would be good to fix them. 

> @eliaslevy Is there a resource on the forward/backward compatibility of Kafka clients/brokers? I was under the impression that only from 1.0.0 onward can you properly use newer clients with older brokers. If that is the case things will get easier for us in the future and we only need the most recent versions.

I believe that @eliaslevy was referring to compatibility other way around: our `0.11` connector can work with `1.0` and `1.1` brokers.

@yanghua if it's problematic, I would be fine with fixing/refactoring kafka 0.10, 0.9 and 0.8 connectors in separate issue (if even fixing them at all). Especially that their code differs quite a bit compared to Kafka 0.11 connector. 

It should be easier to strip current module `flink-connector-kafka-0.11` of kafka dependencies (as @aljoscha proposed, it should be possible and quite easy to do so by wrapping Kafka dependencies into our proxy/bridge interfaces), rename it to `flink-connector-kafka-base-2.0` and implement only `flink-connector-kafka-0.11`, `flink-connector-kafka-1.0`, `flink-connector-kafka-1.1` and `flink-connector-kafka-1.2` by extending from the newly created `flink-connector-kafka-base-2.0`.

**Comment by yanghua** [comment]
@pnowojski Thank you for giving me some tips. I will first pass the current test case, then I will try to refactor `flink-connector-kafka-0.11` and `flink-connector-kafka-1.0`. If I have a question, I will ask you.

**Comment by eliaslevy** [comment]
@aljoscha 
* [KIP-35 - Retrieving protocol version](https://cwiki-test.apache.org/confluence/display/KAFKA/KIP-35+-+Retrieving+protocol+version)
* [KIP-97: Improved Kafka Client RPC Compatibility Policy](https://cwiki-test.apache.org/confluence/display/KAFKA/KIP-97%3A+Improved+Kafka+Client+RPC+Compatibility+Policy)

> This KIP suggests revising the Kafka Java client compatibility policy so that it is two-way, rather than one-way.  Not only should older brokers support newer clients, but newer clients should support older brokers.

> In some cases, older brokers will be unable to perform a feature supported by newer clients.  For example, retrieving message offsets from a message timestamp (KIP-79) is not supported in versions before 0.10.1.  In these cases, the new client will throw an ObsoleteBrokerVersion exception when attempting to perform the operation.  This will let clients know that the reason why the operation cannot be performed is because the broker version in use does not support it.  At the same time, clients which do not attempt to use the new operations will proceed as before.

* [Compatibility Matrix](https://cwiki.apache.org/confluence/display/KAFKA/Compatibility+Matrix)

Basically, any KIP-35 client can talk to a Kafka broker >= 0.10.0, which for the Java client [means](https://cwiki.apache.org/confluence/display/KAFKA/KIP-35+-+Retrieving+protocol+version) versions >= 0.10.2.

@pnowojski I am suggesting that, except possibly for 0.8 and 0.9 clients, there is only a need for a single Kafka connector based on the latest Kafka client (2.0.0 as of now).  There is no need for separate 2.0.0, 1.x.y, 0.11.0.x, and 0.10.x.y connectors. A connector based on 2.0.0 can talk to all those brokers.

**Comment by eliaslevy** [comment]
@yanghua that comment by @pnowojski is not applicable to my question, because I am not suggesting that we upgrade the Kafka dependency without upgrading the connector, nor that we don't support newer Kafka client versions.  I am suggesting we have a single Kafka connector (except for 0.9 and 0.8) that always tracks the latest Kafka client version.

**Comment by yanghua** [comment]
@eliaslevy Things may not be as simple as we are not just using kafka's Public API, but also some non-public APIs and using different Kafka internal fields through reflection. These codes will change with different kafka clients. If we only maintain a connector, then this connector will be filled with the special case handling of the entire kafka client, and if/else statements are everywhere, it may be more difficult to maintain.

Maybe you can look at these [links](https://github.com/apache/flink/pull/6577#issuecomment-414951008) inside.

**Comment by pnowojski** [comment]
@eliaslevy ok, sorry for confusion. In that case my objection is mostly invalid, since we should be able to use the latest kafka client (for example 2.0) to talk with all the previous versions. I expect that the latest version should be also the most stable one.

@yanghua in the option proposed by @eliaslevy we would only be using one version (the latest) of the kafka client, thus there would be no special cases. In this scenario, when we are upgrading from `0.11` to `1.0`, effectively we would drop `0.11` connector code, replacing it with `1.0` code.

Generally speaking I would be in favour of such option. However there would be two questions needing an answer:

1. after upgrading our connector from `0.11` to `1.0`, do we keep old `flink-connector-kafka-0.11` module with small tiny proxy classes redirecting to `flink-connector-kafka-1.0` to provide painless migration?
2. what about testing? Should we still test `flink-connector-kafka-1.0` against `0.11` brokers?

**Comment by yanghua** [comment]
Oh, @eliaslevy  then I understand what you mean, it sounds like a good idea. However, I don't know if Kafka client 0.8 and Kafka client 1.0/2.0 can provide matching or close performance if Kafka Server is version 0.8. 

In addition, I refer to the [source code of Spark](https://github.com/apache/spark/tree/master/external) and find multiple connectors and they are also split by version. I don't mean the choice of Spark is the most suitable, but I think we should consider it carefully before we do this, or we can start a discussion in dev mailing list . 

What do you think? @pnowojski and @aljoscha

**Comment by eliaslevy** [comment]
@yanghua I believe you can't use a modern (>= 0.10.2.0) Kafka client to talk to a 0.8 or 0.9 broker.  That's why I said, we would probably have to keep the 0.8 and 0.9 Kafka connectors, unless we decide we can drop them, if few in the community are still using those versions of Kafka.  They are quite old.  0.8 came out 5 years ago, and 0.9 3 years ago.

But AFAIK there is no performance penalty for using a newer Kafka client (e.g. 2.0.0) with an older KIP-35 supporting broker (e.g. 0.10.2.0).

As for Spark, you will note that they have yet to support any Kafka version greater than 0.10, which are the ones that can be used with any modern client.

@pnowojski rather than releasing `flink-connector-kafka-1.0`, I would just release `flink-connector-kafka` or `flink-connector-kafka-new` or some such, without embedding the Kafka version into the connector artifact id, as it will be updated with every new Kafka version.  And I would drop `flink-connector-kafka-0.11` and `flink-connector-kafka-0.10`.

I would not add a proxy.  It would just add complexity.  If the API doesn't change much, then best for folks to just change their dependencies to point to the new artifact and be done with it.  

Some testing against older brokers would be useful.  Particularly if there are any features we use that may not be available across versions.

**Comment by pnowojski** [comment]
@eliaslevy what do you mean by:
> @pnowojski rather than releasing flink-connector-kafka-1.0, I would just release flink-connector-kafka or flink-connector-kafka-new or some such, without embedding the Kafka version into the connector artifact id, as it will be updated with every new Kafka version.

If we release `flink-connector-kafka` without embedded Kafka version, we can not use reflections for transactional API. To make sure that our `exactly-once` kafka producer work, we must freeze kafka version. For example upgrading from `0.11.0` to `0.11.2` required some tiny change and so does the upgrade to `1.0` (as you can see in this PR).

Another issue is what I have already said before: we can not provide one connector for kafka `0.10` and for `0.11+`, since `0.10` doesn't support transactions, so at the moment we can only provide one connector for versions from `0.11` to `2.0`.

**Comment by eliaslevy** [comment]
@pnowojski I mean that the Flink Kafka connector need not embedded the Kafka version into the connector artifact name.  Obviously, the connector would depend on a specific Kafka client version artifact that would be targeted by the reflection code.

Re: 0.10.  As I mentioned above, you can use a newer client against a 0.10 broker.  If you try to use transactions against the 0.10 broker, the client will raise `UnsupportedVersionException`.   That will let the connector know that the broker does not support transactions.

From the `KafkaProducer` documentation:

> This client can communicate with brokers that are version 0.10.0 or newer. Older or newer brokers 
> may not support certain client features. For instance, the transactional APIs need broker versions 
> 0.11.0 or later. You will receive an UnsupportedVersionException when invoking an API that is not 
> available in the running broker version.

**Comment by eliaslevy** [comment]
If you need to determine that a broker does not support transactions before trying to make use of them, it may be possible to do so via [`NodeApiVersions.latestUsableVersion`](https://github.com/apache/kafka/blob/3cdc78e6bb1f83973a14ce1550fe3874f7348b05/clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java#L91). 

You'd have to iterate over the cluster nodes and check `apiVersions.latestUsableVersion(ApiKeys.PRODUCE)` is `>= 3`.  Version 3 of the produce request [introduced transactions](https://kafka.apache.org/protocol#The_Messages_Produce).

Although deprecated, the Scala [`AdminClient`](https://github.com/apache/kafka/blob/fc5f6b0e46ff81302b3e445fed0cdf454c942792/core/src/main/scala/kafka/admin/AdminClient.scala) has code that demonstrates how to obtain this information.

**Comment by yanghua** [comment]
It sounds like @eliaslevy ‘s idea is valid and feasible.I think we can verify this with kafka 2.0 while keeping the existing connector unchanged.

**Comment by pnowojski** [comment]
I’m not sure if auto detecting support for transactions is worth the trouble. It might be as easy as switching to at least once mode, but it might be just easier to support only Kafka 0.11+ with this unified connector.

**Comment by yanghua** [comment]
@pnowojski What do you think I follow @eliaslevy  advice based on kafka client 2.0 to provide a connector implementation to verify that some discussions are feasible? The existing connector remains unchanged for the time being, what do you think? cc @aljoscha

**Comment by pnowojski** [comment]
I think that's a good plan. Before doing anything fancy, it would be nice to really do a PoC and test the compatibility (upgrade connector to Kafka 2.0 and test it using our tests against Kafka 0.11, 1.0, 1.1, ..., 2.0)

**Comment by yanghua** [comment]
@pnowojski OK, I also claimed [FLINK-9697](https://issues.apache.org/jira/browse/FLINK-9697), which is about Flink kafka connection 2.0. This issue, I am going to try it in the way @eliaslevy  said.

**Comment by aljoscha** [comment]
Responding a bit late but I think @eliaslevy's proposal is definitely the way to go. Thanks a lot for chiming in on this!

And thanks everyone for figuring this out together.

**Comment by yanghua** [comment]
As the PR #6703  is merged and closed, this PR seems to be able to close directly. @aljoscha what do you think?

**Comment by yanghua** [comment]
Since I didn't receive a response, I think it's time for us to close it and focus on the modern kafka connector.
